LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/ConcentratorDeduplication.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace LoRaWan.NetworkServer
+{
+    using LoRaWan.NetworkServer.BasicsStation;
+    using Microsoft.Extensions.Caching.Memory;
+    using Microsoft.Extensions.Logging;
+    using System;
+    using System.Collections.Generic;
+    using System.Security.Cryptography;
+    using System.Text;
+
+    public sealed class ConcentratorDeduplication : IConcentratorDeduplication, IDisposable
+    {
+        internal MemoryCache Cache { get; private set; }
+
+        private readonly ILogger<IConcentratorDeduplication> logger;
+        private static readonly TimeSpan CacheEntryExpiration = TimeSpan.FromMinutes(1);
+
+        public ConcentratorDeduplication(ILogger<IConcentratorDeduplication> logger)
+        {
+            Cache = new MemoryCache(new MemoryCacheOptions());
+            this.logger = logger;
+        }
+
+        public bool ShouldDrop(UpstreamDataFrame updf, StationEui stationEui)
+        {
+            if (updf == null) throw new ArgumentNullException(nameof(updf));
+
+            var key = CreateCacheKey(updf);
+            StationEui previousStation;
+
+            lock (Cache)
+            {
+                if (!Cache.TryGetValue(key, out previousStation))
+                {
+                    AddToCache(key, stationEui);
+                    return false;
+                }
+            }
+
+            if (previousStation == stationEui)
+            {
+                this.logger.LogDebug($"Message received from the same DevAddr: {updf.DevAddr} as before, considered a resubmit.");
+                return false;
+            }
+
+            this.logger.LogInformation($"Duplicate message detected from DevAddr: {updf.DevAddr}, dropping.");
+            return true;
+        }
+
+        internal static string CreateCacheKey(UpstreamDataFrame updf)
+        {
+            using var sha256 = SHA256.Create();
```

Comment:
Do you know how performant is this creation? Would it make sense to keep a reference on the class to avoid recreating it on every request?

Reply:
Just bear in mind, `SHA256` is _not_ thread-safe. If you want to create it once and reuse it, safest is to make it a thread-static via `[ThreadStatic]`. See [C# Cautionary Tale — The dangers of SHA256 reuse](https://wanderingdeveloper.medium.com/c-cautionary-tail-the-dangers-of-sha256-reuse-2b5bb9c6fde9) for background, details and solution.

Reply:
Pfioue looks liek I was wrong, thank you Atif for the hint :)

Reply:
I was debating whether the proposed solution was worth the complexity but at the end [changed to use the ThreadStatic field](c5885a51693f47307a8df5e4999cf66e9c91c75c) .

Reply:
The solution with ThreadStatic backing static field was marked (correctly imo) from CodeQL. 

Using simply a static ThreadStatic member (without a property that checks if it's initialized), doesn't work either from tests since static members are initialized exactly once from the compiler (subsequent ConcentrationDeduplication objects will not have this member initialized)

Reverted to the original code that creates a new instance of Sha256 to avoid this complexity.

Reply:
Reverted back to using ThreadStatic as discussed with Atif.